In [1]:
import pandas as pd
import numpy as np

from sklearn import metrics

import statsmodels.api as sm
from sklearn.model_selection import train_test_split

In [2]:
df = pd.read_csv("BankChurners.csv")

In [3]:
df = df.drop(['CLIENTNUM', 'Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1', 'Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2'], axis=1)

In [4]:
df.drop(df[df['Marital_Status'] == "Unknown"].index, inplace=True)
df.drop(df[df['Income_Category'] == "Unknown"].index, inplace=True)
df.drop(df[df['Education_Level'] == "Unknown"].index, inplace=True)

In [5]:
df['Attrition_Flag'] = df['Attrition_Flag'].replace(['Existing Customer'],'0')
df['Attrition_Flag'] = df['Attrition_Flag'].replace(['Attrited Customer'],'1')

In [6]:
df

Attrition_Flag  Customer_Age Gender  Dependent_count Education_Level  \
0                  0            45      M                3     High School   
1                  0            49      F                5        Graduate   
2                  0            51      M                3        Graduate   
4                  0            40      M                3      Uneducated   
5                  0            44      M                2        Graduate   
...              ...           ...    ...              ...             ...   
10120              0            54      M                1     High School   
10121              0            56      F                1        Graduate   
10122              0            50      M                2        Graduate   
10124              1            44      F                1     High School   
10126              1            43      F                2        Graduate   

      Marital_Status Income_Category Card_Category  Months_on_book  \
0            Married     $60K - $80K          Blue              39   
1             Single  Less than $40K          Blue              44   
2            Married    $80K - $120K          Blue              36   
4            Married     $60K - $80K          Blue              21   
5            Married     $40K - $60K          Blue              36   
...              ...             ...           ...             ...   
10120         Single     $60K - $80K          Blue              34   
10121         Single  Less than $40K          Blue              50   
10122         Single     $40K - $60K          Blue              40   
10124        Married  Less than $40K          Blue              36   
10126        Married  Less than $40K        Silver              25   

       Total_Relationship_Count  Months_Inactive_12_mon  \
0                             5                       1   
1                             6                       1   
2                             4                       1   
4                             5                       1   
5                             3                       1   
...                         ...                     ...   
10120                         5                       2   
10121                         4                       1   
10122                         3                       2   
10124                         5                       3   
10126                         6                       2   

       Contacts_Count_12_mon  Credit_Limit  Total_Revolving_Bal  \
0                          3       12691.0                  777   
1                          2        8256.0                  864   
2                          0        3418.0                    0   
4                          0        4716.0                    0   
5                          2        4010.0                 1247   
...                      ...           ...                  ...   
10120                      0       13940.0                 2109   
10121                      4        3688.0                  606   
10122                      3        4003.0                 1851   
10124                      4        5409.0                    0   
10126                      4       10388.0                 1961   

       Avg_Open_To_Buy  Total_Amt_Chng_Q4_Q1  Total_Trans_Amt  Total_Trans_Ct  \
0              11914.0                 1.335             1144              42   
1               7392.0                 1.541             1291              33   
2               3418.0                 2.594             1887              20   
4               4716.0                 2.175              816              28   
5               2763.0                 1.376             1088              24   
...                ...                   ...              ...             ...   
10120          11831.0                 0.660            15577             114   
10121           3082.0                 0.570            14596     

In [7]:
X = pd.get_dummies(df, columns=['Gender', 'Education_Level', 'Marital_Status', 'Income_Category', 'Card_Category'], drop_first=True)
X = X.drop(['Attrition_Flag'],axis=1)

In [8]:
X_logreg = sm.add_constant(X)

/home/adam_s/um/moje_srod/lib/python3.8/site-packages/statsmodels/tsa/tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


In [9]:
X.columns

Index(['Customer_Age', 'Dependent_count', 'Months_on_book',
       'Total_Relationship_Count', 'Months_Inactive_12_mon',
       'Contacts_Count_12_mon', 'Credit_Limit', 'Total_Revolving_Bal',
       'Avg_Open_To_Buy', 'Total_Amt_Chng_Q4_Q1', 'Total_Trans_Amt',
       'Total_Trans_Ct', 'Total_Ct_Chng_Q4_Q1', 'Avg_Utilization_Ratio',
       'Gender_M', 'Education_Level_Doctorate', 'Education_Level_Graduate',
       'Education_Level_High School', 'Education_Level_Post-Graduate',
       'Education_Level_Uneducated', 'Marital_Status_Married',
       'Marital_Status_Single', 'Income_Category_$40K - $60K',
       'Income_Category_$60K - $80K', 'Income_Category_$80K - $120K',
       'Income_Category_Less than $40K', 'Card_Category_Gold',
       'Card_Category_Platinum', 'Card_Category_Silver'],
      dtype='object')

In [10]:
X_logreg.describe()

const  Customer_Age  Dependent_count  Months_on_book  \
count  7081.0   7081.000000      7081.000000     7081.000000   
mean      1.0     46.347691         2.337805       35.981359   
std       0.0      8.041225         1.291649        8.002609   
min       1.0     26.000000         0.000000       13.000000   
25%       1.0     41.000000         1.000000       31.000000   
50%       1.0     46.000000         2.000000       36.000000   
75%       1.0     52.000000         3.000000       40.000000   
max       1.0     73.000000         5.000000       56.000000   

       Total_Relationship_Count  Months_Inactive_12_mon  \
count               7081.000000             7081.000000   
mean                   3.819376                2.342607   
std                    1.544444                0.995104   
min                    1.000000                0.000000   
25%                    3.000000                2.000000   
50%                    4.000000                2.000000   
75%                    5.000000                3.000000   
max                    6.000000                6.000000   

       Contacts_Count_12_mon  Credit_Limit  Total_Revolving_Bal  \
count            7081.000000   7081.000000          7081.000000   
mean                2.454456   8492.773831          1167.501624   
std                 1.104917   9126.072520           812.315606   
min                 0.000000   1438.300000             0.000000   
25%                 2.000000   2498.000000           463.000000   
50%                 2.000000   4287.000000          1282.000000   
75%                 3.000000  10729.000000          1781.000000   
max                 6.000000  34516.000000          2517.000000   

       Avg_Open_To_Buy  ...  Education_Level_Uneducated  \
count      7081.000000  ...                 7081.000000   
mean       7325.272207  ...                    0.170032   
std        9131.217585  ...                    0.375688   
min           3.000000  ...                    0.000000   
25%        1248.000000  ...                    0.000000   
50%        3250.000000  ...                    0.000000   
75%        9491.000000  ...                    0.000000   
max       34516.000000  ...                    1.000000   

       Marital_Status_Married  Marital_Status_Single  \
count             7081.000000            7081.000000   
mean                 0.503319               0.416325   
std                  0.500024               0.492984   
min                  0.000000               0.000000   
25%                  0.000000               0.000000   
50%                  1.000000               0.000000   
75%                  1.000000               1.000000   
max                  1.000000               1.000000   

       Income_Category_$40K - $60K  Income_Category_$60K - $80K  \
count                  7081.000000                  7081.000000   
mean                      0.199407                     0.155769   
std                       0.399583                     0.362662   
min                       0.000000                     0.000000   
25%                       0.000000                     0.000000   
50%                       0.000000                     0.000000   
75%                       0.000000                     0.000000   
max                       1.000000                     1.000000   

       Income_Category_$80K - $120K  Income_Category_Less than $40K  \
count                    7081.00000                     7081.000000   
mean                        0.16975                        0.394295   
std                         0.37544                        0.488733   
min                         0.00000                        0.000000   
25%                         0.00000                        0.000000   
50%                         0.00000                        0.000000   
75%                         0.00000                        1.000000   
max                         1.00000                        1.000000   

       Card_Ca

In [11]:
y = df['Attrition_Flag']

In [58]:
X_train_logreg, X_test_logreg, y_train_logreg, y_test_logreg = train_test_split(X_logreg, y, random_state=42)

In [14]:
y_train_logreg.astype(float).describe()

count    5310.000000
mean        0.158569
std         0.365308
min         0.000000
25%         0.000000
50%         0.000000
75%         0.000000
max         1.000000
Name: Attrition_Flag, dtype: float64

In [15]:
X_train_logreg.shape

(5310, 30)

In [16]:
y_train_logreg.shape

(5310,)

In [17]:
X_1 = X_train_logreg[[
    'Customer_Age',
    'Dependent_count',
    'Months_on_book',
    'Total_Relationship_Count',
    'Months_Inactive_12_mon',
    'Contacts_Count_12_mon',
    #'Credit_Limit', 
    'Total_Revolving_Bal',
    'Avg_Open_To_Buy',
    'Total_Amt_Chng_Q4_Q1',
    'Total_Trans_Amt',
    'Total_Trans_Ct',
    'Total_Ct_Chng_Q4_Q1',
    'Avg_Utilization_Ratio',
    'Gender_M',
    'Education_Level_High School',
    'Education_Level_Graduate',
    'Education_Level_Uneducated',
    #'Education_Level_College',
    'Education_Level_Post-Graduate',
    'Education_Level_Doctorate',
    'Marital_Status_Married',
    'Marital_Status_Single',
    #'Marital_Status_Divorced',
    'Income_Category_$60K - $80K',
    'Income_Category_Less than $40K',
    'Income_Category_$80K - $120K',
    'Income_Category_$40K - $60K',
    #'Income_Category_$120K +',
    'Card_Category_Gold',
    'Card_Category_Silver',
    'Card_Category_Platinum'
               ]]
X_1 = sm.add_constant(X_1)

/home/adam_s/um/moje_srod/lib/python3.8/site-packages/statsmodels/tsa/tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


In [18]:
X_1

const  Customer_Age  Dependent_count  Months_on_book  \
8101     1.0            46                4              36   
8916     1.0            51                1              37   
7945     1.0            45                4              36   
10061    1.0            44                2              31   
4743     1.0            48                3              42   
...      ...           ...              ...             ...   
5242     1.0            49                3              36   
7365     1.0            44                5              30   
7413     1.0            50                1              36   
7681     1.0            49                4              35   
1158     1.0            26                0              15   

       Total_Relationship_Count  Months_Inactive_12_mon  \
8101                          2                       1   
8916                          1                       3   
7945                          1                       3   
10061                         3                       2   
4743                          3                       2   
...                         ...                     ...   
5242                          5                       3   
7365                          3                       3   
7413                          4                       3   
7681                          5                       3   
1158                          5                       1   

       Contacts_Count_12_mon  Total_Revolving_Bal  Avg_Open_To_Buy  \
8101                       1                 1043           1128.0   
8916                       2                 2416          24902.0   
7945                       3                    0           4688.0   
10061                      3                 2028           2154.0   
4743                       4                    0          19075.0   
...                      ...                  ...              ...   
5242                       4                    0           1438.3   
7365                       1                 1726            983.0   
7413                       2                    0           2317.0   
7681                       2                 2133            905.0   
1158                       2                 1720           6547.0   

       Total_Amt_Chng_Q4_Q1  ...  Education_Level_Doctorate  \
8101                  0.700  ...                          0   
8916                  0.917  ...                          0   
7945                  0.593  ...                          0   
10061                 0.658  ...                          0   
4743                  0.807  ...                          0   
...                     ...  ...                        ...   
5242                  0.929  ...                          0   
7365                  0.698  ...                          0   
7413                  0.734  ...                          0   
7681                  0.689  ...                          0   
1158                  0.633  ...                          0   

       Marital_Status_Married  Marital_Status_Single  \
8101                        1                      0   
8916                        0                      1   
7945                        0                      1   
10061                       0                      1   
4743                        0                      1   
...                       ...                    ...   
5242                        1                      0   
7365                        0                      1   
7413                        0                      1   
7681                        1                      0   
1158                        0                      1   

       Income_Category_$60K - $80K  Income_Category_Less than $40K  \
8101                             0                               0   
8916                             0                               0   
7945                             0                          

Credit_Limit powoduje że macierzy nie da się odwrócić (wyznacznik równy 0). Reszta zmiennych nie sprawia więkrzych problemów. Kwestia teraz zbudowania odpowiedniego modelu + interpretacje + ilorazy szans.

In [19]:
log_reg_1 = sm.Logit(y_train_logreg.astype(float), X_1.astype(float)).fit()

Optimization terminated successfully.
         Current function value: 0.237683
         Iterations 8


In [20]:
print(log_reg_1.summary())

                           Logit Regression Results                           
Dep. Variable:         Attrition_Flag   No. Observations:                 5310
Model:                          Logit   Df Residuals:                     5281
Method:                           MLE   Df Model:                           28
Date:                Wed, 09 Feb 2022   Pseudo R-squ.:                  0.4565
Time:                        20:41:36   Log-Likelihood:                -1262.1
converged:                       True   LL-Null:                       -2322.0
Covariance Type:            nonrobust   LLR p-value:                     0.000
                                     coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
const                              7.2959      0.631     11.567      0.000       6.060       8.532
Customer_Age                      -0.0134      0.011     -1.259      0.

In [21]:
# test łącznej istotności Education_Level

x_vars = log_reg_1.summary2().tables[1].index
#x_vars
wald_str = ' = '.join(list(x_vars[15:20]))
wald_str

'Education_Level_High School = Education_Level_Graduate = Education_Level_Uneducated = Education_Level_Post-Graduate = Education_Level_Doctorate'

In [22]:
wald_test = log_reg_1.wald_test(wald_str) # Wald test
wald_test

<class 'statsmodels.stats.contrast.ContrastResults'>
<Wald test (chi2): statistic=[[8.99697749]], p-value=0.061175073835732924, df_denom=4>

Można usunąć zmienne dotyczące wykształcenia, pojedyncze zmienne nie istotne statystycznie, również łączny test to potwierdza (p-value > 0.05).

**Pominięcie Education_Level**

In [23]:
X_2 = X_train_logreg[[
    'Customer_Age',
    'Dependent_count',
    'Months_on_book',
    'Total_Relationship_Count',
    'Months_Inactive_12_mon',
    'Contacts_Count_12_mon',
    #'Credit_Limit', 
    'Total_Revolving_Bal',
    'Avg_Open_To_Buy',
    'Total_Amt_Chng_Q4_Q1',
    'Total_Trans_Amt',
    'Total_Trans_Ct',
    'Total_Ct_Chng_Q4_Q1',
    'Avg_Utilization_Ratio',
    'Gender_M',
    #'Education_Level_High School',
    #'Education_Level_Graduate',
    #'Education_Level_Uneducated',
    #'Education_Level_College',
    #'Education_Level_Post-Graduate',
    #'Education_Level_Doctorate',
    'Marital_Status_Married',
    'Marital_Status_Single',
    #'Marital_Status_Divorced',
    'Income_Category_$60K - $80K',
    'Income_Category_Less than $40K',
    'Income_Category_$80K - $120K',
    'Income_Category_$40K - $60K',
    #'Income_Category_$120K +',
    'Card_Category_Gold',
    'Card_Category_Silver',
    'Card_Category_Platinum'
               ]]
X_2 = sm.add_constant(X_2)

/home/adam_s/um/moje_srod/lib/python3.8/site-packages/statsmodels/tsa/tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


In [24]:
log_reg_2 = sm.Logit(y_train_logreg.astype(float), X_2.astype(float)).fit()

Optimization terminated successfully.
         Current function value: 0.238503
         Iterations 8


In [25]:
print(log_reg_2.summary())

                           Logit Regression Results                           
Dep. Variable:         Attrition_Flag   No. Observations:                 5310
Model:                          Logit   Df Residuals:                     5286
Method:                           MLE   Df Model:                           23
Date:                Wed, 09 Feb 2022   Pseudo R-squ.:                  0.4546
Time:                        20:41:36   Log-Likelihood:                -1266.5
converged:                       True   LL-Null:                       -2322.0
Covariance Type:            nonrobust   LLR p-value:                     0.000
                                     coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
const                              7.2979      0.615     11.865      0.000       6.092       8.503
Customer_Age                      -0.0131      0.011     -1.245      0.

In [26]:
# test łącznej istotności Marital_Status
x_vars = log_reg_2.summary2().tables[1].index
#x_vars
wald_str = ' = '.join(list(x_vars[15:17]))
wald_str


'Marital_Status_Married = Marital_Status_Single'

In [27]:
wald_test = log_reg_2.wald_test(wald_str) # joint test
wald_test

<class 'statsmodels.stats.contrast.ContrastResults'>
<Wald test (chi2): statistic=[[32.0434897]], p-value=1.5075947447532242e-08, df_denom=1>

Ten sam test na zmienną Marital_Status, w tym przypadku można zostawić obie zmienne (p-value < 0.05).

**Pominięcie zmiennej Months_on_book**

In [28]:
X_3 = X_train_logreg[[
    'Customer_Age',
    'Dependent_count',
    #'Months_on_book',
    'Total_Relationship_Count',
    'Months_Inactive_12_mon',
    'Contacts_Count_12_mon',
    #'Credit_Limit', 
    'Total_Revolving_Bal',
    'Avg_Open_To_Buy',
    'Total_Amt_Chng_Q4_Q1',
    'Total_Trans_Amt',
    'Total_Trans_Ct',
    'Total_Ct_Chng_Q4_Q1',
    'Avg_Utilization_Ratio',
    'Gender_M',
    #'Education_Level_High School',
    #'Education_Level_Graduate',
    #'Education_Level_Uneducated',
    #'Education_Level_College',
    #'Education_Level_Post-Graduate',
    #'Education_Level_Doctorate',
    'Marital_Status_Married',
    'Marital_Status_Single',
    #'Marital_Status_Divorced',
    'Income_Category_$60K - $80K',
    'Income_Category_Less than $40K',
    'Income_Category_$80K - $120K',
    'Income_Category_$40K - $60K',
    #'Income_Category_$120K +',
    'Card_Category_Gold',
    'Card_Category_Silver',
    'Card_Category_Platinum'
               ]]
X_3 = sm.add_constant(X_3)

/home/adam_s/um/moje_srod/lib/python3.8/site-packages/statsmodels/tsa/tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


In [29]:
log_reg_3 = sm.Logit(y_train_logreg.astype(float), X_3.astype(float)).fit()

Optimization terminated successfully.
         Current function value: 0.238549
         Iterations 8


In [30]:
print(log_reg_3.summary())

                           Logit Regression Results                           
Dep. Variable:         Attrition_Flag   No. Observations:                 5310
Model:                          Logit   Df Residuals:                     5287
Method:                           MLE   Df Model:                           22
Date:                Wed, 09 Feb 2022   Pseudo R-squ.:                  0.4545
Time:                        20:41:36   Log-Likelihood:                -1266.7
converged:                       True   LL-Null:                       -2322.0
Covariance Type:            nonrobust   LLR p-value:                     0.000
                                     coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
const                              7.3161      0.615     11.901      0.000       6.111       8.521
Customer_Age                      -0.0190      0.006     -2.977      0.

**Pominięcie Avg_Utilization_Ratio**

In [31]:
X_4 = X_train_logreg[[
    'Customer_Age',
    'Dependent_count',
    #'Months_on_book',
    'Total_Relationship_Count',
    'Months_Inactive_12_mon',
    'Contacts_Count_12_mon',
    #'Credit_Limit', 
    'Total_Revolving_Bal',
    'Avg_Open_To_Buy',
    'Total_Amt_Chng_Q4_Q1',
    'Total_Trans_Amt',
    'Total_Trans_Ct',
    'Total_Ct_Chng_Q4_Q1',
    #'Avg_Utilization_Ratio',
    'Gender_M',
    #'Education_Level_High School',
    #'Education_Level_Graduate',
    #'Education_Level_Uneducated',
    #'Education_Level_College',
    #'Education_Level_Post-Graduate',
    #'Education_Level_Doctorate',
    'Marital_Status_Married',
    'Marital_Status_Single',
    #'Marital_Status_Divorced',
    'Income_Category_$60K - $80K',
    'Income_Category_Less than $40K',
    'Income_Category_$80K - $120K',
    'Income_Category_$40K - $60K',
    #'Income_Category_$120K +',
    'Card_Category_Gold',
    'Card_Category_Silver',
    'Card_Category_Platinum'
                    ]]
X_4 = sm.add_constant(X_4)

/home/adam_s/um/moje_srod/lib/python3.8/site-packages/statsmodels/tsa/tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


In [32]:
log_reg_4 = sm.Logit(y_train_logreg.astype(float), X_4.astype(float)).fit()

Optimization terminated successfully.
         Current function value: 0.238666
         Iterations 8


In [33]:
print(log_reg_4.summary2())

                                Results: Logit
Model:                   Logit                 Pseudo R-squared:      0.454    
Dependent Variable:      Attrition_Flag        AIC:                   2578.6337
Date:                    2022-02-09 20:41      BIC:                   2723.3353
No. Observations:        5310                  Log-Likelihood:        -1267.3  
Df Model:                21                    LL-Null:               -2322.0  
Df Residuals:            5288                  LLR p-value:           0.0000   
Converged:               1.0000                Scale:                 1.0000   
No. Iterations:          8.0000                                                
-------------------------------------------------------------------------------
                                Coef.  Std.Err.    z     P>|z|   [0.025  0.975]
-------------------------------------------------------------------------------
const                           7.2940   0.6144  11.8722 0.0000  6.0899  

In [34]:
# test łącznej istotności Income_Category

x_vars = log_reg_4.summary2().tables[1].index
#x_vars
wald_str = ' = '.join(list(x_vars[15:19]))
wald_str


'Income_Category_$60K - $80K = Income_Category_Less than $40K = Income_Category_$80K - $120K = Income_Category_$40K - $60K'

In [35]:
wald_test = log_reg_4.wald_test(wald_str) # Wald test
wald_test

<class 'statsmodels.stats.contrast.ContrastResults'>
<Wald test (chi2): statistic=[[12.70740372]], p-value=0.0053140468223303, df_denom=3>

Test łącznej istotności parametrów przy zmiennych związanych z dochodem. Parametry przy zmiennych Income_Category_$60K - $80K oraz Income_Category_$80K - $120K nie są istotne statystycznie, natomiast pozostałe dwie kategorie oraz wspólny test wskazują na istotności statystyczną parametrów. Zasadne jest pozostawienie zmiennych w modelu.

In [36]:
# test łącznej istotności Card_Category

x_vars = log_reg_4.summary2().tables[1].index
#x_vars
wald_str = ' = '.join(list(x_vars[19:22]))
wald_str


'Card_Category_Gold = Card_Category_Silver = Card_Category_Platinum'

In [37]:
wald_test = log_reg_4.wald_test(wald_str) # Wald test
wald_test

<class 'statsmodels.stats.contrast.ContrastResults'>
<Wald test (chi2): statistic=[[1.70146941]], p-value=0.42710102424866614, df_denom=2>

Wszystkie parametry przy zmiennych dotyczących rodzaju karty są nie istotne statystycznie na poziomie istotności 0,05. Potwierdza to również test łączny dla tych parametrów. Zasadnym jest pominięcie tych zmiennych w modelu.

In [38]:
# test łącznej istotności Marital_Status

x_vars = log_reg_4.summary2().tables[1].index
#x_vars
wald_str = ' = '.join(list(x_vars[13:15]))
wald_str


'Marital_Status_Married = Marital_Status_Single'

In [39]:
wald_test = log_reg_4.wald_test(wald_str) # Wald test
wald_test

<class 'statsmodels.stats.contrast.ContrastResults'>
<Wald test (chi2): statistic=[[31.64116951]], p-value=1.8545596792041255e-08, df_denom=1>

Łączny test parametrów przy zmiennych dotyczących stanu cywilnego wskazuje na istotność statystyczną.

In [40]:
X_5 = X_train_logreg[[
    'Customer_Age',
    'Dependent_count',
    #'Months_on_book',
    'Total_Relationship_Count',
    'Months_Inactive_12_mon',
    'Contacts_Count_12_mon',
    #'Credit_Limit', 
    'Total_Revolving_Bal',
    'Avg_Open_To_Buy',
    'Total_Amt_Chng_Q4_Q1',
    'Total_Trans_Amt',
    'Total_Trans_Ct',
    'Total_Ct_Chng_Q4_Q1',
    #'Avg_Utilization_Ratio',
    'Gender_M',
    #'Education_Level_High School',
    #'Education_Level_Graduate',
    #'Education_Level_Uneducated',
    #'Education_Level_College',
    #'Education_Level_Post-Graduate',
    #'Education_Level_Doctorate',
    'Marital_Status_Married',
    'Marital_Status_Single',
    #'Marital_Status_Divorced',
    'Income_Category_$60K - $80K',
    'Income_Category_Less than $40K',
    'Income_Category_$80K - $120K',
    'Income_Category_$40K - $60K'
    #'Income_Category_$120K +',
    #'Card_Category_Gold',
    #'Card_Category_Silver',
    #'Card_Category_Platinum'
                    ]]
X_5 = sm.add_constant(X_5)

/home/adam_s/um/moje_srod/lib/python3.8/site-packages/statsmodels/tsa/tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


In [41]:
log_reg_5 = sm.Logit(y_train_logreg.astype(float), X_5.astype(float)).fit()

Optimization terminated successfully.
         Current function value: 0.239254
         Iterations 8


In [42]:
print(log_reg_5.summary2())

                                Results: Logit
Model:                   Logit                 Pseudo R-squared:      0.453    
Dependent Variable:      Attrition_Flag        AIC:                   2578.8797
Date:                    2022-02-09 20:41      BIC:                   2703.8493
No. Observations:        5310                  Log-Likelihood:        -1270.4  
Df Model:                18                    LL-Null:               -2322.0  
Df Residuals:            5291                  LLR p-value:           0.0000   
Converged:               1.0000                Scale:                 1.0000   
No. Iterations:          8.0000                                                
-------------------------------------------------------------------------------
                                Coef.  Std.Err.    z     P>|z|   [0.025  0.975]
-------------------------------------------------------------------------------
const                           7.1118   0.6083  11.6908 0.0000  5.9195  

**Pominięcie zmiennej Avg_Open_To_Buy**

In [43]:
X_6 = X_train_logreg[[
    'Customer_Age',
    'Dependent_count',
    #'Months_on_book',
    'Total_Relationship_Count',
    'Months_Inactive_12_mon',
    'Contacts_Count_12_mon',
    #'Credit_Limit', 
    'Total_Revolving_Bal',
    #'Avg_Open_To_Buy',
    'Total_Amt_Chng_Q4_Q1',
    'Total_Trans_Amt',
    'Total_Trans_Ct',
    'Total_Ct_Chng_Q4_Q1',
    #'Avg_Utilization_Ratio',
    'Gender_M',
    #'Education_Level_High School',
    #'Education_Level_Graduate',
    #'Education_Level_Uneducated',
    #'Education_Level_College',
    #'Education_Level_Post-Graduate',
    #'Education_Level_Doctorate',
    'Marital_Status_Married',
    'Marital_Status_Single',
    #'Marital_Status_Divorced',
    'Income_Category_$60K - $80K',
    'Income_Category_Less than $40K',
    'Income_Category_$80K - $120K',
    'Income_Category_$40K - $60K'
    #'Income_Category_$120K +',
    #'Card_Category_Gold',
    #'Card_Category_Silver',
    #'Card_Category_Platinum'
                    ]]
X_6 = sm.add_constant(X_6)

/home/adam_s/um/moje_srod/lib/python3.8/site-packages/statsmodels/tsa/tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


In [44]:
log_reg_6 = sm.Logit(y_train_logreg.astype(float), X_6.astype(float)).fit()

Optimization terminated successfully.
         Current function value: 0.239339
         Iterations 8


In [45]:
print(log_reg_6.summary2())

                                Results: Logit
Model:                   Logit                 Pseudo R-squared:      0.453    
Dependent Variable:      Attrition_Flag        AIC:                   2577.7852
Date:                    2022-02-09 20:41      BIC:                   2696.1774
No. Observations:        5310                  Log-Likelihood:        -1270.9  
Df Model:                17                    LL-Null:               -2322.0  
Df Residuals:            5292                  LLR p-value:           0.0000   
Converged:               1.0000                Scale:                 1.0000   
No. Iterations:          8.0000                                                
-------------------------------------------------------------------------------
                                Coef.  Std.Err.    z     P>|z|   [0.025  0.975]
-------------------------------------------------------------------------------
const                           7.0106   0.5987  11.7093 0.0000  5.8371  

In [46]:
# test łącznej istotności Marital_Status

x_vars = log_reg_6.summary2().tables[1].index
#x_vars
wald_str = ' = '.join(list(x_vars[12:14]))
wald_str


'Marital_Status_Married = Marital_Status_Single'

In [47]:
wald_test = log_reg_6.wald_test(wald_str) # Wald test
wald_test

<class 'statsmodels.stats.contrast.ContrastResults'>
<Wald test (chi2): statistic=[[30.99708796]], p-value=2.5841585610824837e-08, df_denom=1>

Łączny test parametrów przy zmiennych dotyczących stanu cywilnego wskazuje na istotność statystyczną.

In [48]:
# test łącznej istotności Income_Category

x_vars = log_reg_6.summary2().tables[1].index
#x_vars
wald_str = ' = '.join(list(x_vars[14:18]))
wald_str


'Income_Category_$60K - $80K = Income_Category_Less than $40K = Income_Category_$80K - $120K = Income_Category_$40K - $60K'

In [49]:
wald_test = log_reg_6.wald_test(wald_str) # Wald test
wald_test

<class 'statsmodels.stats.contrast.ContrastResults'>
<Wald test (chi2): statistic=[[9.22302688]], p-value=0.026467986954468205, df_denom=3>

Test łącznej istotności parametrów przy zmiennych związanych z dochodem. Parametry przy zmiennych Income_Category_$60K - $80K oraz Income_Category_$80K - $120K nie są istotne statystycznie, natomiast pozostałe dwie kategorie oraz wspólny test wskazują na istotności statystyczną parametrów. Zasadne jest pozostawienie zmiennych w modelu.

**Finalny model to log_reg_6**

**Ilorazy szans**

In [51]:
np.exp(log_reg_6.params)

const                             1108.339307
Customer_Age                         0.981015
Dependent_count                      1.132214
Total_Relationship_Count             0.656484
Months_Inactive_12_mon               1.632241
Contacts_Count_12_mon                1.620977
Total_Revolving_Bal                  0.999051
Total_Amt_Chng_Q4_Q1                 0.596824
Total_Trans_Amt                      1.000491
Total_Trans_Ct                       0.888688
Total_Ct_Chng_Q4_Q1                  0.069245
Gender_M                             0.380292
Marital_Status_Married               0.519896
Marital_Status_Single                0.967759
Income_Category_$60K - $80K          0.724283
Income_Category_Less than $40K       0.563193
Income_Category_$80K - $120K         0.928912
Income_Category_$40K - $60K          0.494123
dtype: float64

Wzrost wieku o jeden rok spowoduje spadek szans na rezygnację z usług kart kredytowych o 2%.
Kolejna dodatkowa osoba na utrzymaniu klienta spowoduje wzrost szans na rezygnację z usług kart kredytowych o 13%.
Osoby w związku małżeńskim, w porównaniu do osób rozwiedzionych, mają o 52% mniejsze szanse na rezygnację z usług kart kredytowych. Osoby o dochodach poniżej 40 tys. dolarów mają o 44% mniejsze szanse na rezygnację z usług kart kredytowych, w porównaniu do osób o dochodach większych niż 120 tys. dolarów. Mężczyźni mają o 62% mniejsze szanse na rezygnację z usług kart kredytowych w porównaniu do kobiet.

In [52]:
X_6_test = X_test_logreg[[
    'Customer_Age',
    'Dependent_count',
    #'Months_on_book',
    'Total_Relationship_Count',
    'Months_Inactive_12_mon',
    'Contacts_Count_12_mon',
    #'Credit_Limit', 
    'Total_Revolving_Bal',
    #'Avg_Open_To_Buy',
    'Total_Amt_Chng_Q4_Q1',
    'Total_Trans_Amt',
    'Total_Trans_Ct',
    'Total_Ct_Chng_Q4_Q1',
    #'Avg_Utilization_Ratio',
    'Gender_M',
    #'Education_Level_High School',
    #'Education_Level_Graduate',
    #'Education_Level_Uneducated',
    #'Education_Level_College',
    #'Education_Level_Post-Graduate',
    #'Education_Level_Doctorate',
    'Marital_Status_Married',
    'Marital_Status_Single',
    #'Marital_Status_Divorced',
    'Income_Category_$60K - $80K',
    'Income_Category_Less than $40K',
    'Income_Category_$80K - $120K',
    'Income_Category_$40K - $60K'
    #'Income_Category_$120K +',
    #'Card_Category_Gold',
    #'Card_Category_Silver',
    #'Card_Category_Platinum'
               ]]
X_6_test = sm.add_constant(X_6_test)

/home/adam_s/um/moje_srod/lib/python3.8/site-packages/statsmodels/tsa/tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


In [53]:
y_test_logreg = y_test_logreg.astype(float)
y_test_logreg.describe()

count    1771.000000
mean        0.153021
std         0.360109
min         0.000000
25%         0.000000
50%         0.000000
75%         0.000000
max         1.000000
Name: Attrition_Flag, dtype: float64

In [54]:
predicted_y_logreg = log_reg_6.predict(X_6_test)
predicted_y_logreg = np.round(predicted_y_logreg, 0).astype(float)
predicted_y_logreg

9866    0.0
9239    0.0
963     0.0
1999    1.0
2531    0.0
       ... 
9288    0.0
1961    0.0
6059    0.0
9300    0.0
2827    0.0
Length: 1771, dtype: float64

In [55]:
print(metrics.accuracy_score(y_test_logreg, predicted_y_logreg))
print(metrics.confusion_matrix(y_test_logreg, predicted_y_logreg))
print(metrics.classification_report(y_test_logreg, predicted_y_logreg))

0.9040090344438171
[[1443   57]
 [ 113  158]]
              precision    recall  f1-score   support

         0.0       0.93      0.96      0.94      1500
         1.0       0.73      0.58      0.65       271

    accuracy                           0.90      1771
   macro avg       0.83      0.77      0.80      1771
weighted avg       0.90      0.90      0.90      1771



Accuracy na zbiorze testowym wynosi 90%. Model 158 razy poprawnie przewidział rezygnację z usług kart kredytowych, gdzie w rzeczywistości również miało to miejsce. 1443 razy model poprawnie przewidział dalsze korzystanie z usług kart kredytowych, gdzie w rzeczywistości również miało to miejsce. 113 razy model przewidział dalsze korzystanie z usług kart kredytowych, gdzie w rzeczywistości nie miało to miejsca. 57 razy model przewidział rezygnację z usług kart kredytowych, gdzie w rzeczywistości nie miało to miejsca.

In [56]:
pred_train_y = log_reg_6.predict(X_6)
pred_train_y = np.round(pred_train_y, 0).astype(float)
pred_train_y

8101     0.0
8916     0.0
7945     0.0
10061    0.0
4743     1.0
        ... 
5242     0.0
7365     0.0
7413     1.0
7681     0.0
1158     0.0
Length: 5310, dtype: float64

In [57]:
print(metrics.accuracy_score(y_train_logreg.astype(float), pred_train_y))
print(metrics.confusion_matrix(y_train_logreg.astype(float), pred_train_y))
print(metrics.classification_report(y_train_logreg.astype(float), pred_train_y))

0.9024482109227872
[[4314  154]
 [ 364  478]]
              precision    recall  f1-score   support

         0.0       0.92      0.97      0.94      4468
         1.0       0.76      0.57      0.65       842

    accuracy                           0.90      5310
   macro avg       0.84      0.77      0.80      5310
weighted avg       0.90      0.90      0.90      5310



Prezycja modelu na zbiorze treningowym